File: go/quantum_safe_kem.go
// Quantum Safe - Key Encapsulation Mechanisms (KEM)
package pqc

import (
	"github.com/cloudflare/circl/kem/kyber/kyber768"
	"github.com/cloudflare/circl/kem/kyber/kyber1024"
	"github.com/cloudflare/circl/kem/frodo/frodo640shake"
	"github.com/open-quantum-safe/liboqs-go/oqs"
)

func MLKem768Encapsulation() {
	pk, sk, _ := kyber768.GenerateKeyPair(nil)
	ct, ss, _ := kyber768.Encapsulate(pk)
	ssDecap, _ := kyber768.Decapsulate(sk, ct)
	_ = ssDecap
}

func MLKem1024Example() {
	publicKey, secretKey, _ := kyber1024.GenerateKeyPair(nil)
	ciphertext, sharedSecret, _ := kyber1024.Encapsulate(publicKey)
	decrypted, _ := kyber1024.Decapsulate(secretKey, ciphertext)
	_ = decrypted
	_ = sharedSecret
}

func FrodoKEMExample() {
	pk, sk, _ := frodo640shake.GenerateKeyPair(nil)
	ct, ss, _ := frodo640shake.Encapsulate(pk)
	_ = ss
	_ = ct
	_ = sk
}

func OqsMLKem512() {
	kem := oqs.KeyEncapsulation{}
	kem.Init("Kyber512", nil)
	pk, _ := kem.GenerateKeyPair()
	ct, ss, _ := kem.Encapsulate(pk)
	_ = ct
	_ = ss
}

func KyberKeyGeneration() {
	scheme := kyber768.Scheme()
	pk, sk, _ := scheme.GenerateKeyPair()
	_ = pk
	_ = sk
}

File: rust/quantum_safe_kem.rs
// Quantum Safe - Key Encapsulation Mechanisms (KEM)
use pqcrypto_kyber::kyber768;
use pqcrypto_kyber::kyber1024;
use pqcrypto_frodo::frodokem640shake;
use oqs::kem::{Kem, Algorithm};

fn ml_kem_768_encapsulation() {
    let (pk, sk) = kyber768::keypair();
    let (ss, ct) = kyber768::encapsulate(&pk);
    let ss_dec = kyber768::decapsulate(&ct, &sk);
}

fn ml_kem_1024_example() {
    let (public_key, secret_key) = kyber1024::keypair();
    let (shared_secret, ciphertext) = kyber1024::encapsulate(&public_key);
    let decrypted = kyber1024::decapsulate(&ciphertext, &secret_key);
}

fn frodokem_example() {
    let (pk, sk) = frodokem640shake::keypair();
    let (ss, ct) = frodokem640shake::encapsulate(&pk);
}

fn oqs_ml_kem_512() {
    let kem = Kem::new(Algorithm::Kyber512).unwrap();
    let (pk, sk) = kem.keypair().unwrap();
    let (ct, ss) = kem.encapsulate(&pk).unwrap();
}
